test/functional/test-element-stub.js

```diff
@@ -0,0 +1,156 @@
+ * limitations under the License.
+ */
+
+import {ElementStub, resetAdScriptInsertedOrPresent, calculateAdScriptUrl,}
+  from '../../src/element-stub';
+import {createIframePromise} from '../../testing/iframe';
+import {setModeForTesting, getMode} from '../../src/mode';
+
+
+describe('test-element-stub', () => {
+
+  let iframe;
+
+  function getElementStub(name) {
+    return createIframePromise().then(f => {
+      iframe = f;
+      testElement = iframe.doc.createElement(name);
+      testElement.setAttribute('width', '300');
+      testElement.setAttribute('height', '250');
+      testElement.setAttribute('type', 'a9');
+      testElement.setAttribute('data-aax_size', '300*250');
+      testElement.setAttribute('data-aax_pubname', 'abc123');
+      testElement.setAttribute('data-aax_src', '302');
+      const link = iframe.doc.createElement('link');
+      link.setAttribute('rel', 'canonical');
+      link.setAttribute('href', 'blah');
+      iframe.doc.head.appendChild(link);
+      return iframe.addElement(testElement);
+    });
+  }
+
+  it('insert script', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInsertedOrPresent();
+      expect(iframe.doc.querySelectorAll('amp-ad').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      elementStub = new ElementStub(iframe.doc.body.querySelector('#parent').firstChild);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .not.to.be.null;
+    });
+  });
+
+  it('only insert script once', () => {
+    return getElementStub('amp-ad').then(() => {
+      resetAdScriptInsertedOrPresent();
+      expect(iframe.doc.querySelectorAll('amp-ad').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      elementStub = new ElementStub(iframe.doc.body.querySelector('#parent').firstChild);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+      elementStub = new ElementStub(iframe.doc.body.querySelector('#parent').firstChild);
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
+        .length).to.equal(1);
+    });
+  });
+
+  it('insert script for amp-embed', () => {
+    return getElementStub('amp-embed').then(() => {
+      resetAdScriptInsertedOrPresent();
+      expect(iframe.doc.querySelectorAll('amp-embed').length).to.equal(1);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-ad"]'))
+        .to.be.null;
+      expect(iframe.doc.head.querySelector('[custom-element="amp-embed"]'))
+        .to.be.null;
+      elementStub = new ElementStub(iframe.doc.body.querySelector('#parent').firstChild);
+      expect(iframe.doc.head.querySelector('[custom-element="amp-embed"]'))
+        .to.be.null;
+      expect(iframe.doc.head.querySelectorAll('[custom-element="amp-ad"]')
```

Comment:
`expect(…).to.have.length(…)`

That gets you much better errors when it fails.